spartan_parallel/src/sparse_mlpoly.rs

```diff
@@ -1,98 +1,89 @@
 #![allow(clippy::type_complexity)]
 #![allow(clippy::too_many_arguments)]
 #![allow(clippy::needless_range_loop)]
+use crate::scalar::SpartanExtensionField;
+
 use super::dense_mlpoly::DensePolynomial;
-use super::dense_mlpoly::{
-  EqPolynomial, IdentityPolynomial, PolyCommitment, PolyCommitmentGens, PolyEvalProof,
-};
+use super::dense_mlpoly::{EqPolynomial, IdentityPolynomial, PolyEvalProof};
 use super::errors::ProofVerifyError;
 use super::math::Math;
 use super::product_tree::{DotProductCircuit, ProductCircuit, ProductCircuitEvalProofBatched};
 use super::random::RandomTape;
-use super::scalar::Scalar;
 use super::timer::Timer;
 use super::transcript::{AppendToTranscript, ProofTranscript};
 use core::cmp::Ordering;
 use merlin::Transcript;
 use serde::{Deserialize, Serialize};
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct SparseMatEntry {
+pub struct SparseMatEntry<S: SpartanExtensionField> {
   row: usize,
   col: usize,
-  val: Scalar,
+  val: S,
 }
 
-impl SparseMatEntry {
-  pub fn new(row: usize, col: usize, val: Scalar) -> Self {
+impl<S: SpartanExtensionField> SparseMatEntry<S> {
+  pub fn new(row: usize, col: usize, val: S) -> Self {
     SparseMatEntry { row, col, val }
   }
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct SparseMatPolynomial {
+pub struct SparseMatPolynomial<S: SpartanExtensionField> {
   num_vars_x: usize,
   num_vars_y: usize,
-  M: Vec<SparseMatEntry>,
-}
-
-pub struct Derefs {
-  row_ops_val: Vec<DensePolynomial>,
-  col_ops_val: Vec<DensePolynomial>,
-  comb: DensePolynomial,
+  M: Vec<SparseMatEntry<S>>,
 }
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct DerefsCommitment {
-  comm_ops_val: PolyCommitment,
+pub struct Derefs<S: SpartanExtensionField> {
+  row_ops_val: Vec<DensePolynomial<S>>,
+  col_ops_val: Vec<DensePolynomial<S>>,
+  comb: DensePolynomial<S>,
 }
 
-impl Derefs {
-  pub fn new(row_ops_val: Vec<DensePolynomial>, col_ops_val: Vec<DensePolynomial>) -> Self {
+impl<S: SpartanExtensionField> Derefs<S> {
+  pub fn new(row_ops_val: Vec<DensePolynomial<S>>, col_ops_val: Vec<DensePolynomial<S>>) -> Self {
     assert_eq!(row_ops_val.len(), col_ops_val.len());
 
+    let ret_row_ops_val = row_ops_val.clone();
+    let ret_col_ops_val = col_ops_val.clone();
+
     let derefs = {
       // combine all polynomials into a single polynomial (used below to produce a single commitment)
-      let comb = DensePolynomial::merge(row_ops_val.iter().chain(col_ops_val.iter()));
+      let comb = DensePolynomial::merge(row_ops_val.into_iter().chain(col_ops_val.into_iter()));
```

Comment:
You could just clone here, then you don't need to change all the other lines.  Makes the diff easier to review.